### DISCUSSION TITLE
After migration to 0.19.0, CssStyle is forcing place items attribute

### DISCUSSION THREAD

**Discussion by sunildhiman90** [varabyte/kobweb#592] [Q&A]
AFter i migrated my simple project to kobweb 0.19.0 and kotlin to 2.0.10
using ` ./gradlew kobwebMigrateToK2 `
and `../gradlew kobwebMigrateToCssStyle`

Then my some styles are forcefully adding this class while im inspecting the html in browser
`.kobweb-box.kobweb-align-top-start`
and even after i added 
```
.placeItems(
                alignItems = com.varabyte.kobweb.compose.css.AlignItems.Center,
                justifyItems = JustifyItems.Center
            )
```
But still its using that one forcefully and not allowing me to override that by .placeItems in CssStyle

How to fix this issue? Thnx

If relevant, specify:
* OS: mac
* Kobweb libs version:  0.19.0

**Comment by bitspittle** [comment]
Check the alignment parameter

https://github.com/varabyte/kobweb/blob/514b883d8b00c485352481d3248f70eea05b8b6e/frontend/kobweb-compose/src/jsMain/kotlin/com/varabyte/kobweb/compose/foundation/layout/Box.kt#L43

And set it to `FromStyle`:

https://github.com/varabyte/kobweb/blob/514b883d8b00c485352481d3248f70eea05b8b6e/frontend/kobweb-compose/src/jsMain/kotlin/com/varabyte/kobweb/compose/ui/Alignment.kt#L47

Alternatively, use a Div and not a Box. Let me know if neither of those approaches worked.

**Reply by sunildhiman90** [reply]
UPDATE:
Yes actually, i already tried alignment just after creating the issue and it worked.
Adding contentAlignment to Box fixed the issue, 
`contentAlignment = Alignment.Center,`

But I have one question here, It creates some confusion sometimes, If on jetpack compose like composables like row, column and box, Should we direclty use modifiers and their arrangement properties,Or CssStlye.toModifier?

I mean sometimes we are not able to align things correctly with simple modifiers and arrangements, then we need to use CssStyle.
If im not wrong, CssStyle is mainly for html element composables like div etc. Otherwise we can directly use modifiers, right?

But still not sure where i was wrong, sometime i was not able to align things properly with just modifiers and arrangements on Row, Column, Box etc earlier.  so i used CssStye, and it creates confusion sometime which one should i prefer. 

**Anyway, May be i need to explore it more.**

By the way, Thanx @bitspittle , this framework is very nice for Android/Kotlin Developers.
Im waiting for SSR to be implemented in this, Then it will just become too awesome.
At the moment, i use nextjs for websites whenever needed, But once this will be ready with SSR, it will be the thing to go. Boom!

btw, do we have any progress on SSR and Hydration things?
Actually i was checking the tickets on SSR and hydration and there is no update there, thats why i wanted to know its status on SSR and hydration.
Thanku.

**Reply by bitspittle** [reply]
In general, if parameters are exposed by Kobweb / Silk widgets, use them directly. Otherwise, you can follow guidance here about styling, although sometimes it is more art than science: https://github.com/varabyte/kobweb?tab=readme-ov-file#html-styling

There is no plan to prioritize SSR anytime soon. Doing great SSR is extremely hard because 1) it forces you to use a fullstack solution, which is generally slower and more expensive to host and 2) makes things like locally interactive widgets much harder to write (since you have to keep that state in sync with the server usually?) See also [this blog post section]( https://dev.to/hillaryprosper_wahua_bce/server-side-rendering-ssr-in-web-development-1oi0#:~:text=Challenges%20and%20Limitations%20of%20Server%2DSide%20Rendering%20(SSR)).

The current export process works well enough for most (all?) cases that users have brought up to us so far.

That said, if you're really dead set on SSR and  haven't heard of it already, https://github.com/rjaros/kilua is a project that is building up a web framework from the ground up with SSR support.

**Reply by sunildhiman90** [reply]
Thanks @bitspittle for getting time to reply. 

As you said current process works for most cases. So i want to know is there any way to do seo for dynamic routes like blog article detail, ecommerce product detail etc In kobweb? not sure if may be im missing anything about kobweb?

And that is the reason i was asking about SSR thing. 

And as far as about kilua, yes i checked that out. Thats great for SSR. But few things kobweb provides some composable components rows, columns, etc similar to compose that i was liking here, they were more easier to use and similar to compose than directly using html and css things. And also it was similar to nextjs as well and easier page setups and other things, thats why i liked it more.

**Reply by bitspittle** [reply]
I am honestly not too sure how to do effective SEO for dynamic routes in general. If you want to do it for just a handful of URLs, you can at least [request explicit exports for those specific routes](https://github.com/varabyte/kobweb?tab=readme-ov-file#dynamic-routes-and-exporting).

There may other approaches as well (maybe a dynamically generated or carefully handcrafted sitemap?). [Here's an article](https://medium.com/@rameshkannanyt0078/next-js-is-an-excellent-framework-for-building-server-rendered-or-statically-exported-react-apps-04b875eff6cf) discussing the problem for nextjs users.

**Reply by sunildhiman90** [reply]
Ok great. Thanks for clearing this out. I was in confusion that nextjs has SSR and some other feature and that would be enough for SEO for dynamic routes. BUT may be that too requires generating sitemaps dynamically. 
And Can we have something similar in kobweb here, generating sitemaps dynamically and then somehow setup that in kobweb, the way they are specifying in next.config.js ? Any trick or suggestion here may be?

**Reply by bitspittle** [reply]
I do want to support sitemap generation eventually. But current priorities are the 1.0 roadmap (you can find a link in the README of curious, just search for "roadmap") which will have me busy for I think another 6-12 months. In the meanwhile, people can definitely do sitemap stuff themselves and won't be blocked by the feature being missing.

**Reply by sunildhiman90** [reply]
Ok Thnx, and i understand you will be busy with that roadmap. One last question, as u said people might be doing this stuff themselves, so do u have any reference or anyone who have done the similar thing becoz that can save me some time atleast , otherwise i will see whenever gets some time, i will explore that topic in reference of kobweb. And also being the author of this framework, do you have some suggestion for me if i try to do it myself sometime. Any tips to start with his process., how to proceed?

The way we are generating static site, that same concept can be helpful here i guess for dynamic routes( looping over the list of articles). But not sure, as you have better understanding of the project.

**Reply by bitspittle** [reply]
I think for broader questions about if anyone has any experience or not, you may want to check out the Discord and/or Kotlin Slack channel (both of which are linked to from the README).

Very close to the top of the README links to some tutorial videos. Unfortunately, as they are older, a few concepts might be slightly out of date (mainly, if you ever see `by ComponentStyle` that is now `= CssStyle`); but overall they should be really useful in helping you get a good mental model for the framework. 

Between those videos and reading through the beginner section of the README, that should really cover most fundamental concepts. It's honestly not too bad -- you declare pages and then you declare layouts and CSS styles, that's like 99% of it.

We are in the process of improving documentation for new users now, including looking into creating official YT videos, but this will easily be a many month long process.

For now, best advice is to use example sites (run `kobweb create` and you'll see a handful of simple examples to learn from; also, you can find showcased projects in the Discord and Slack channels which are often open sourced).

**Reply by sunildhiman90** [reply]
Thanks @bitspittle , I am already in those communities, and i have checked the templates of kobweb as well, But i was just concerned of the generating sitemap for dynamic routes. That was the only query from me as of now.  ANyway thanks, i will check those groups and channels asking for this sitemap thing. Thanx for your help and suggestions at all.

**Reply by bitspittle** [reply]
You may be breaking new ground! If you find a solution that you like, please feel welcome to share it here if you remember.

You may also find a lot of luck putting code into your gradle build script. Check out https://github.com/varabyte/kobweb?tab=readme-ov-file#generating-site-code-at-compile-time if you haven't seen it already.

**Reply by sunildhiman90** [reply]
Haha, yes thats true. If i get that what im looking for it will be just awesome. And thanks that script will be helpful offcourse for generating code at compile time.   

And the thing is may be i will be looking into it in bits and pieces due to job plus other part time stuff. But i would definitely like to give that thing a hand whenever will be getting  some free time slots. 

And dont worry if i face any problems while doing so, i will be posting updates here asking you for the help.

Anyway the way we are generating static site from the code,  where is the script for that  logic ?
